src/get_time.cpp

```diff
@@ -31,32 +33,121 @@ namespace cass {
                     static_cast<uint64_t>(ft.dwLowDateTime)) -
                    116444736000000000LL; // 100 nanosecond increments between
                                          // Jan. 1, 1601 - Jan. 1, 1970
-  return ns100 / 10000;                  // 100 nanoseconds to milliseconds
+  return ns100 / 10;                     // 100 nanosecond increments to microseconds
+}
+
+uint64_t get_time_monotonic_ns() {
+  if (ClockInfo::frequency() != 0) {
+    LARGE_INTEGER counter;
+    QueryPerformanceCounter(&counter);
+
+    counter.QuadPart *= NANOSECONDS_PER_SECOND;
+    return static_cast<uint64_t>(counter.QuadPart) / ClockInfo::frequency();
+  } else {
+    return get_time_since_epoch_us() * NANOSECONDS_PER_MICROSECOND;
+  }
 }
 
 #elif defined(__APPLE__) && defined(__MACH__)
 
-uint64_t get_time_since_epoch_ms() {
+// Information on converting the absolute time to nanaoseconds can be found
+// here: https://developer.apple.com/library/content/qa/qa1398/_index.html.
+
+class ClockInfo {
+public:
+  ClockInfo() {
+    mach_timebase_info_data_t info;
+    mach_timebase_info(&info);
+    frequency_ = info.numer / info.denom;
+  }
+
+  static uint64_t frequency() { return frequency_; }
+
+private:
+  static uint64_t frequency_;
+};
+
+uint64_t ClockInfo::frequency_;
+
+static ClockInfo __clock_info__; // Initializer
+
+uint64_t get_time_since_epoch_us() {
   struct timeval tv;
   gettimeofday(&tv, NULL);
-  return static_cast<uint64_t>(tv.tv_sec)  * 1000 +
-         static_cast<uint64_t>(tv.tv_usec) / 1000;
+  return static_cast<uint64_t>(tv.tv_sec) * 1000000 +
+         static_cast<uint64_t>(tv.tv_usec);
+}
+
+uint64_t get_time_monotonic_ns() {
+  uint64_t time = mach_absolute_time();
+  return time * ClockInfo::frequency();
 }
 
 #else
 
-uint64_t get_time_since_epoch_ms() {
+class ClockInfo {
+public:
+  ClockInfo() {
+    struct timespec res;
+    struct timespec tp;
+    supports_monotonic_ = clock_getres(CLOCK_MONOTONIC, &res) == 0 &&
+                          clock_gettime(CLOCK_MONOTONIC, &tp) == 0;
+  }
+
+  static bool supports_monotonic() { return supports_monotonic_; }
+
+private:
+  static bool supports_monotonic_;
+};
+
+bool ClockInfo::supports_monotonic_;
+
+static ClockInfo __clock_info__; // Initializer
+
+
+uint64_t get_time_since_epoch_us() {
   struct timespec ts;
   clock_gettime(CLOCK_REALTIME, &ts);
-  return static_cast<uint64_t>(ts.tv_sec)  * 1000 +
-         static_cast<uint64_t>(ts.tv_nsec) / 1000000;
+  return static_cast<uint64_t>(ts.tv_sec)  * 1000000 +
+         static_cast<uint64_t>(ts.tv_nsec) / 1000;
+}
+
+uint64_t get_time_monotonic_ns() {
+  if (ClockInfo::supports_monotonic()) {
+    struct timespec tp;
+    clock_gettime(CLOCK_MONOTONIC, &tp);
+    return static_cast<uint64_t>(tp.tv_sec) * NANOSECONDS_PER_SECOND +
+           static_cast<uint64_t>(tp.tv_nsec);
+  } else {
+    return get_time_since_epoch_us() * NANOSECONDS_PER_MICROSECOND;
+  }
```

Comment:
I'm torn on one point here -- this whole file is platform-specific, so I feel like it would be better to split it into three files (say get_time-win.cpp, get_time-mac.cpp, get_time-unix.cpp) to make it easier for a developer to look at in the future. On the other hand, it's 150+ lines of code, which isn't that big.

Consider splitting, but I'm fine either way.